### PR TITLE
Customer search field shows results for all words

### LINKED ISSUE #22652: BO ORDER CREATE - Customer search field shows results for all words
#### Describe the bug

When you go to create a new order from BO and you search for customer, it shows results for all words you enter.
You search for John Doe and instead of one or few John Does, it finds all Johns and all Doe. There are billion of results... 

#### Expected behavior

It should search only for the phrase as a whole.

#### Steps to reproduce

1. **BO**>**SELL**>**Customer**
2. Create two customers: "John Smith" and "Sam Doe"
3. **BO**>**SELL**>**Orders**
4. Create Order
5. make a search with "John Doe"
6. See error > there's 3 customers diplayed

#### Additional information

* PrestaShop version: 1.7.6.2
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | When you go to create a new order from BO and you search for customer, it shows results for all words you enter. You search for John Doe and instead of one or few John Does, it finds all Johns and all Do
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/22652
| How to test?      | Please indicate how to best verify that this PR is correct.
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

This PR add an option to allow the merchant to choose the type of search.
This option is very useful for shops with hundreds of thousands of customers.

Capture d’écran de 2021-10-05 08-49-22
Capture d’écran de 2021-10-05 08-49-45
ScreenShot Tool -20211005094229

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26109)

### CODE REVIEW

**Comment by Hlavtox** [comment]
@okom3pom Awesome!!! I did it locally but with this switch it's gonna be even better. 🚀

What about putting it to Order settings section, and maybe creating a new card called "Manual order"?

Výstřižek

**Comment by ghost** [comment]
@Hlavtox 

This is a good idea! 

To not change the default behaviour we could change the words of the switch? And reverse it ?

Thank you for your idea and wording is welcome my english is poor

**Comment by Hlavtox** [comment]
@okom3pom I think we should insert the default value to configuration anyway, so it should not matter what the default value is. ;-)

Just put the default setting here:
https://github.com/PrestaShop/PrestaShop/blob/b4e28291ebf5bc34ee6b70095f0e414a76915f18/install-dev/upgrade/sql/1.7.9.0.sql#L8

**Comment by Hlavtox** [comment]
@okom3pom And put the default value also to the install. https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/data/xml/configuration.xml

I would keep the current behavior for upgrading users, but set the better one for new installs. WDYT?

**Comment by ghost** [comment]
I will do this, but for 1.7.9.0.sql, I need to port the change to the autoupgrade module.

https://github.com/PrestaShop/PrestaShop/pull/26084#pullrequestreview-771048306

**Comment by ghost** [comment]
_I would keep the current behavior for upgrading users, but set the better one for new installs. WDYT?_

The better one :)

**Comment by Hlavtox** [comment]
Aaaaaha, didn't know this, sorry 🚀

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: Maybe we should add a placeholder on the page and use it as a separator. Here we are choosing the white space as the one and only 'separator' character that can be used.

Maybe some people would expect the separator character to be ; or , or / . We just need to make sure there is no confusion, it's not ambiguous.

**Review by matks — COMMENTED** [review]
Hello @okom3pom what a nice idea !

So the logic is overall good, but as said by @Hlavtox 
- we need to think about what happens for people who install (is it enabled by default)
- we need to think about what happens for people who upgrade (is it enabled by default)

Finally, there is one little thing I am afraid: the confusion of putting this new option inside the BO. We need to be sure that it is really really clear that "this option controls how the block Search for Customers behaves internally in the Create an Order Back Office page". We dont want people to turn this thing on and off and expect something else to happen.

So basically we need @PrestaShop/product-team and @Julievrz help to make sure this feature is used correctly by users who dot no see the PR context.

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: @matks The "space" separator is just the original behavior that just sucks... 😄😄 I would just change it to searching the whole word, but @okom3pom tried to be nice and keep the option of original behavior, if anyone wanted to do it. I would not invest any time developing this feature.

I think all people will change to our new behavior right away, I search for Mathieu Ferment and it will show me all Mathieus and all Ferments, completely unusable.

Comment (Hlavtox) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: My proposal - original behavior for people who upgrade, new behavior for new users.

**Comment by ghost** [comment]
Hi @matks,

If the PM is ok I would use @Hlavtox's method with the option on a new card in the order configuration page with the title: Manual order.

So no ambiguity :)

The option is to keep the default behaviour. In a shop like mine with 350k customers it's horrible the default behaviour.

**Comment by Julievrz** [comment]
Hi,
Why not display the text directly on the Orders > Create order page? Something like 
"Display exact matches only" or "Display all results"?

**Review by matthieu-rolland — COMMENTED** [review]
That's a great idea ! 

I wouldn't aprove this PR without having the product team's feedback though @marionf @MatShir 

Code wise it seems good but we also need to check where this option would be implemented, which could imply changes in your implementation

**Comment by Hlavtox** [comment]
@Julievrz It would only occupy space Julie, merchants will configure this one time and then forget about it. I think we could just even change the behavior and nobody would mind.

**Comment by matks** [comment]
> I wouldn't approve this PR without having the product team's feedback though @marionf @MatShir

So I add the "Waiting for PM" label, I ping @PrestaShop/product-team so that they can tell us if the feature is good and well built in their opinion

**Comment by MatShir** [comment]
IMO, the OR search does not make sense here, you are looking for a specific person and not a multiple results. The "OR" would be interesting if the user would search for many results and find the best, one. So the option adding the parameter  just for manual order might be too much. But If we still want to add a configuration option, why not use the "OR" into the search bar like many advanced search bar, so the user can still use it whitout going into the parameter (or guessing that there is a paramenter for it)

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]:
```suggestion
        $phrases = $configuration->get('PS_CUSTOMER_SEARCH_FULL') ? [$query] : explode(' ', $query);
```

A small clarification of code. After that, it's an approval.

**Review by Progi1984 — COMMENTED** [review]
@okom3pom Thank you for your contribution. A small feedback for code.

After that, we need @Julievrz for wording as the feature is already "PM :heavy_check_mark:"

**Comment by MatShir** [comment]
@Progi1984, I didn't approve the feature yet, I'm waiting for @okom3pom inputs. If we follow my suggestion, there shouldn't have any wording needed

**Comment by Progi1984** [comment]
@MatShir But in the linked issue : 
image

**Comment by MatShir** [comment]
Yes but the issue is very different from the behavior of the PR

**Comment by Hlavtox** [comment]
@okom3pom Man just hardcode it. Nobody in his right mind would want the original behavior. :D

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php]: Feedback applied

**Review by matks — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @okom3pom 

LGTM, QA :heavy_check_mark: 
https://watch.screencastify.com/v/LAaSHigVFIOhQ7HyFOra

Thanks!

**Comment by Progi1984** [comment]
We need a wording validation : Ping @Julievrz or @PrestaShop/product-team

**Comment by MatShir** [comment]
The PR still not fixes the issue, the option part is not required ! As @Hlavtox said, we can integrate directly or add a "OR" search advanced method. 
There is no point having the parameter, like I said before. I was waiting for @okom3pom response with a "waiting for author" but was removed. 🤷 
So the PR QA was approved, @okom3pom do you feel like change the behavior?

**Review by PierreRambaud — COMMENTED** [review]
I think this is something we should add in fixture data. (`configuration.xml`)  :thinking:

**Comment by Hlavtox** [comment]
@okom3pom Pls reverse the behavior man, it should be “no or” by default. So if there is a switch, it should be called “Enable OR behavior in search”.

**Comment by Julievrz** [comment]
> @okom3pom Pls reverse the behavior man, it should be “no or” by default. So if there is a switch, it should be called “Enable OR behavior in search”.

If I understand well @MatShir's comment, there should be no wording at all. If users want to use the "OR" feature, they should be able to enter "OR" in the search bar, as they would do on Google. So I'm not sure to understand why there are 2 new strings in this PR.

**Comment by ghost** [comment]
I'll work on it next week. Thanks for not closing the PR

**Comment by MatShir** [comment]
thanks, @okom3pom, we appreciate :)

**Comment by ghost** [comment]
Closed by https://github.com/PrestaShop/PrestaShop/pull/27441

**Pull request closed without merging**
